File: data/src/main/java/home/oleg/placesnearme/repositories/VenueRepositoryImpl.java
package home.oleg.placesnearme.repositories;

import java.util.List;
import java.util.Map;

import home.oleg.placenearme.models.Venue;
import home.oleg.placenearme.repositories.Category;
import home.oleg.placenearme.repositories.VenueRepository;
import home.oleg.placesnearme.service.IFourSquareAPI;
import io.reactivex.Single;

public class VenueRepositoryImpl implements VenueRepository {

    private final IFourSquareAPI api;
    private final QueryParamCreator queryParamCreator;

    public VenueRepositoryImpl(IFourSquareAPI api, QueryParamCreator queryParamCreator) {
        this.api = api;
        this.queryParamCreator = queryParamCreator;
    }

    @Override
    public Single<List<Venue>> getRecommendedByCategory(Category category, Filter filter) {
        Map<String, String> queryMap = queryParamCreator.create(category, filter);

        return api.explore(queryMap).map(fullResponse -> fullResponse.getResponse().getVenues());
    }

    @Override
    public Single<List<Venue>> search(String query, Filter filter) {
        Map<String, String> queryMap = queryParamCreator.create(query, filter);

        return api.search(queryMap).map(fullResponse -> fullResponse.getResponse().getVenues());
    }

}

File: data/src/main/java/home/oleg/placesnearme/data/repositories/VenueRepositoryImpl.kt
package home.oleg.placesnearme.data.repositories

import home.oleg.placenearme.models.Section
import home.oleg.placenearme.models.Venue
import home.oleg.placenearme.repositories.VenueRepository
import home.oleg.placenearme.repositories.VenueRequestParams
import home.oleg.placesnearme.data.mapper.VenueMapper
import home.oleg.placesnearme.network.service.IFourSquareAPI
import io.reactivex.Single

class VenueRepositoryImpl(private val api: IFourSquareAPI) : VenueRepository {
    private val queryParamCreator = QueryParamCreator()

    override fun getRecommendedBySection(section: Section, requestParams: VenueRequestParams): Single<List<Venue>> {
        val queryMap = queryParamCreator.create(section, requestParams)

        return api.explore(queryMap)
                .map { it.response }
                .map { it.venues }
                .map { VenueMapper.map(it) }
    }

    override fun search(query: String, requestParams: VenueRequestParams): Single<List<Venue>> {
        val queryMap = queryParamCreator.create(query, requestParams)
        return api.search(queryMap)
                .map { it.response }
                .map { it.venues }
                .map { VenueMapper.map(it) }
    }

}
